File: 3_linear_data_structures/ch07/two_sum3_test.py
import unittest  # The test framework
from typing import List

# https://leetcode.com/problems/two-sum/


def two_sum(nums: List[int], target: int) -> List[int]:
    nums_map = {}
    # 키와 값을 바꿔서 딕셔너리로 저장
    for i, num in enumerate(nums):
        nums_map[num] = i

    # 타겟에서 첫 번째 수를 뺀 결과를 키로 조회
    for i, num in enumerate(nums):
        if target - num in nums_map and i != nums_map[target - num]:
            return [i, nums_map[target - num]]


class TestTwoSum3(unittest.TestCase):
    def test_case1(self):
        # given
        nums = [2, 7, 11, 15]
        target = 9
        # when
        res = two_sum(nums, target)
        # then
        assert res == [0, 1]

    def test_case2(self):
        # given
        nums = [3, 2, 4]
        target = 6
        # when
        res = two_sum(nums, target)
        # then
        assert res == [1, 2]

    def test_case3(self):
        # given
        nums = [3, 3]
        target = 6
        # when
        res = two_sum(nums, target)
        # then
        assert res == [0, 1]


if __name__ == "__main__":
    unittest.main()

File: 3_linear_data_structures/ch07/two-sum3.test.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
const twoSum = (nums, target) => {
  const numMap = new Map();
  nums.forEach((number, index) => {
    numMap.set(number, index);
  });
  for (const [index, number] of nums.entries()) {
    if (numMap.has(target - number) && numMap.get(target - number) !== index) {
      return [index, numMap.get(target - number)];
    }
  }
};

describe(`${twoSum.name}`, () => {
  it('case1', () => {
    // given

    const nums = [2, 7, 11, 15];
    const target = 9;
    // when
    const res = twoSum(nums, target);
    // then
    expect(res).toEqual([0, 1]);
  });

  it('case2', () => {
    // given
    const nums = [3, 2, 4];
    const target = 6;
    // when
    const res = twoSum(nums, target);
    // then
    expect(res).toEqual([1, 2]);
  });

  it('case3', () => {
    // given
    const nums = [3, 3];
    const target = 6;
    // when
    const res = twoSum(nums, target);
    // then
    expect(res).toEqual([0, 1]);
  });
});
